NDTensors/src/lib/Sectors/src/category_product.jl

```diff
@@ -34,78 +45,206 @@ function Base.show(io::IO, s::CategoryProduct)
+recover_key(T::Type, c::CategoryProduct) = recover_key(T, categories(c))
+function recover_key(T::Type, fused::Tuple)
+  # here fused contains at leat one GradedUnitRange
+  g0 = reduce(×, fused)
+  # convention: keep unsorted blocklabels as produced by F order loops in ×
+  new_labels = recover_key.(T, GradedAxes.blocklabels(g0))
+  new_blocklengths =
+    LabelledNumbers.labelled.(GradedAxes.unlabel.(BlockArrays.blocklengths(g0)), new_labels)
+  return GradedAxes.gradedrange(new_blocklengths)
+end
+
+sector(T::Type{<:CategoryProduct}, cats::Tuple) = sector(categories_type(T), cats)
+sector(T::Type, cats::Tuple) = sector(T(cats))  # recover NamedTuple
+sector(args...; kws...) = CategoryProduct(args...; kws...)
+
+# =================================  Cartesian Product  ====================================
 ×(c1::AbstractCategory, c2::AbstractCategory) = ×(CategoryProduct(c1), CategoryProduct(c2))
 function ×(p1::CategoryProduct, p2::CategoryProduct)
   return CategoryProduct(categories_product(categories(p1), categories(p2)))
 end
 
-categories_product(l1::NamedTuple, l2::NamedTuple) = union_keys(l1, l2)
-
-categories_product(l1::Tuple, l2::Tuple) = (l1..., l2...)
-
+×(a, g::AbstractUnitRange) = ×(to_graded_axis(a), g)
+×(g::AbstractUnitRange, b) = ×(g, to_graded_axis(b))
 ×(nt1::NamedTuple, nt2::NamedTuple) = ×(CategoryProduct(nt1), CategoryProduct(nt2))
 ×(c1::NamedTuple, c2::AbstractCategory) = ×(CategoryProduct(c1), CategoryProduct(c2))
 ×(c1::AbstractCategory, c2::NamedTuple) = ×(CategoryProduct(c1), CategoryProduct(c2))
 
-#
-# Dictionary-like implementation
-#
+function ×(l1::LabelledNumbers.LabelledInteger, l2::LabelledNumbers.LabelledInteger)
+  c3 = LabelledNumbers.label(l1) × LabelledNumbers.label(l2)
+  m3 = LabelledNumbers.unlabel(l1) * LabelledNumbers.unlabel(l2)
+  return LabelledNumbers.labelled(m3, c3)
+end
+
+function ×(g1::AbstractUnitRange, g2::AbstractUnitRange)
+  v = map(
+    ((l1, l2),) -> l1 × l2,
+    Iterators.flatten((
+      Iterators.product(BlockArrays.blocklengths(g1), BlockArrays.blocklengths(g2)),
+    ),),
+  )
+  return GradedAxes.gradedrange(v)
+end
+
+# ====================================  Fusion rules  ======================================
+# generic case: fusion returns a GradedAxes, even for fusion with Empty
+function fusion_rule(::SymmetryStyle, s1::CategoryProduct, s2::CategoryProduct)
+  return to_graded_axis(categories_fusion_rule(categories(s1), categories(s2)))
+end
+
+# Abelian case: fusion returns CategoryProduct
+function fusion_rule(::AbelianGroup, s1::CategoryProduct, s2::CategoryProduct)
+  return categories_fusion_rule(categories(s1), categories(s2))
+end
+
+# Empty case
+function fusion_rule(
+  ::EmptyCategory, ::CategoryProduct{Tuple{}}, ::CategoryProduct{Tuple{}}
+)
+  return sector()
+end
+
+# EmptyCategory acts as trivial on any AbstractCategory, not just CategoryProduct
+function fusion_rule(::SymmetryStyle, ::CategoryProduct{Tuple{}}, c::AbstractCategory)
```

Comment:
What about the case of `NamedTuple{}`?

Also, are these definitions with empty `CategoryProduct` really needed? Can't those corner cases be handled by the more generic definitions above? I.e. is this primarily an optimization, or done this way because you think it simplifies the code design in some way or is challenging to handle with a generic definition?

Reply:
Having two different kinds of `EmptyCategory` was a hassle, so I removed the `NamedTuple{}` one. It is replaced by `Tuple{}` at init, no more need to duplicate code.

Currently it not possible to fuse two different `AbstractCategory` unless both of them are `CategoryProduct`. These lines allow for one exception, fusing `sector()` with any `AbstractCategory`, as `sector()` may arise when fusing an empty tuple of `GradedUnitRange`.

Reply:
Couldn't you define a type alias `const EmptyCategoryProduct{Categories<:Union{Tuple{},NamedTuple{}} = CategoryProduct{Categories}` to handle both cases? I think using `EmptyCategoryProduct` in these kinds of function definitions would be clearer/more readable anyway, even if we still primarily prefer `Tuple{}` as the default empty category type.

I'm not really following your comment above about fusing two different `AbstractCategory`, why couldn't you fuse two non-product `AbstractCategory`s? I'm also still not sure why we need special definitions for empty category products and why those couldn't be handled by more generic definitions.

Reply:
@emstoudenmire I think you were involved in writing some of this `CategoryProduct` code originally, I would be curious to hear your thoughts on the design of this code as well.

Reply:
Adding an `EmptyCategoryProduct` alias is a good idea.

For your second point, it seems to me this is not about these two lines but about the generic design and permissiveness. The question is what should be the output of `U1(1) ⊗ SU2(1)`? Currently, this gives an error, as would `sector(U1(1)) ⊗ sector(SU2(1))`. If you want to get a category product you need either `U1(1) × SU2(1)` or `sector(A=U1(1)) ⊗ sector(B=SU2(1))`.

These lines allow for `sector() ⊗ U1(1)` to return `U1(1)`, which is NOT a `CategoryProduct`. So even if we were to define `U1(1) ⊗ SU2(1)` to be `U1(1) × SU2(1)`, you would need a dedicated code for `EmptyCategoryProduct`.

Reply:
I see, thanks for the clarification, that helps frame the design choices that we are deciding on here, and the inspiration for the current definitions.

I agree that we shouldn't allow `U1(1) ⊗ SU2(1)`, it wouldn't make sense for that to be `U1(1) × SU2(1)` since that imposes some arbitrary assumption about which "slot" those symmetries sectors are in, i.e. I picture `U1(1)` and `SU2(1)` to be equivalent to `CategoryProduct((U1(1),))` and `CategoryProduct((SU2(1),))`, so both symmetries are in the first "slot".

I guess another way to get `U1(1) × SU2(1)` from `⊗` could be something like `(U1(1) × TrivialCategory()) ⊗ (TrivialCategory() × SU2(1))`, where `TrivialCategory()` is some universal zero/trivial sector (what we call `QN()` right now in ITensors.jl).

So I see your point about maybe defining `sector() ⊗ U1(1)` as `U1(1)`. However, I find `sector(...)` to be kind of a weird thing to begin with, see my comment: https://github.com/ITensor/ITensors.jl/pull/1363#discussion_r1774362911. This may be even more reason to not have a `sector` function, since it seems to serve too many purposes (i.e. either it is `CategoryProduct`, or it is some universal identity category when it is used as `sector()`). Instead of `sector() ⊗ U1(1)`, I would prefer to be more specific and have something like `TrivialCategory()` (maybe there is a better name for that, please feel free to suggest one), so that is spelled like `TrivialCategory() ⊗ U1(1) === U1(1)`.

However, there is still a question about what `CategoryProduct() ⊗ U1(1)` does, probably it _has_ to be defined as `U1(1)` because of the way other products act, for example it is a natural consequence of rules like `CategoryProduct((; A=U1(1))) ⊗ CategoryProduct(; B=SU2(1))) == CategoryProduct((; A=U1(1), B=SU2(1)))`. That does come back to my point that the code that implements that logic in principle should cover the empty category product limit, so I don't see why we need specialized definitions for the empty category product limit.

Reply:
`sector` was designed to construct a `CategoryProduct` from a tuple or a `NamedTuple` (and it can be generalized). As said below, this function may be replace by directly calling `CategoryProduct`, this is not the point here. The question is what should `CategoryProduct()` output? Without any constraint in `CategoryProduct` input, it creates a `CategoryProduct{NamedTuple{}}`, and `CategoryProduct(())` creates a `CategoryProduct{Tuple{}}`. It is natural to expect these objects to act as trivial.

Independently from these implementations details, it is convenient to have a trivial sector to be used in other components, as your `QN()`. We previously agreed to merge these two concepts. I later found it impractical to have two versions of it, so I made `sector()` to return `sector(())`, but we could very well have a custom `EmptyCategoryType <: AbstractCategory` or an `EmptyCategoryProduct <: CategoryProduct` for this. I think `CategoryProduct()` and `CategoryProduct(())` should still return that object.

Finally, there should be a way to get `CategoryProduct() ⊗ U1(1) = U1(1)`. As this is an exception to the rule "fusing two different categories that are not `CategoryProduct` is forbidden", I believe there has to be dedicated code lines. The code may be slightly different if `EmptyCategoryType` is not a `CategoryProduct`, that is the main change I can think of within this main picture.

Reply:
Why is there a rule that you shouldn't fuse two categories that aren't `CategoryProduct`? What about `CategoryProduct((U1(2),) ⊗ U1(1)`? That seems like valid code, and could return `CategoryProduct((U1(3),)`, for example.

Reply:
I'm fine with defining `CategoryProduct()` as `CategoryProduct(())`, by the way.

Let me be more specific about my suggestion. I would suggest defining:
```julia
function fusion_rule(style::SymmetryStyle, c1::CategoryProduct, c2::AbstractCategory)
  return fusion_rule(style, c1, CategoryProduct(c2))
end
```
where `CategoryProduct(c::AbstractCategory) = CategoryProduct((c,))`. Then, the implementation can be handled by generic fusion rules for two `CategoryProduct`s.

Reply:
That definition would implicitly cover the case of empty categories, without requiring specialized definitions.

Reply:
@ogauthe I'm not sure what conclusion we came to here, I'm still hoping that we can minimize the number of `fusion_rule` definitions that hardcode `EmptyCategoryProduct` as inputs and instead rely on the generic `fusion_rule` definition, which if written correctly should handle the empty category product limit without added complexity.

Reply:
The current plan (unless there is some objection which I haven't heard yet) is to introduce a new `struct TrivialSector <: AbstractCategory end` type that acts as a universal trivial sector/irrep, and not treat `CategoryProduct{Tuple{}}` or `CategoryProduct{NamedTuple{()}}` as special in any way and not have specialized code for those types. I think that addresses the issues being discussed here in a simple way.